lib/AST/Decl.cpp

```diff
@@ -482,12 +486,49 @@ case DeclKind::ID: return cast<ID##Decl>(this)->getLocFromSource();
   llvm_unreachable("Unknown decl kind");
 }
 
-SourceLoc Decl::getLoc() const {
+Decl::CachedExternalSourceLocs*
+Decl::calculateSerializedLocs() const {
+  auto *Result = getASTContext().Allocate<Decl::CachedExternalSourceLocs>();
+  auto *File = cast<FileUnit>(getDeclContext()->getModuleScopeContext());
+  auto Locs = File->getBasicLocsForDecl(this);
+  if (!Locs.hasValue()) {
+    return Result;
+  }
```

Comment:
One trick you can do here for the common case of "no locs" is to have them all share a single `static const Decl::CachedExternalSourceLocs NoLocs;`. That avoids an allocation.

Reply:
Good point! Fixed in the newest commit.

Reply:
Ah, I suggest *not* doing it in Decl.h; there's no reason to expose it outside of this function. The `const` is nice too because then it can go in the const data section. You don't actually need the *struct* to be mutable if it's stored out-of-line.